### PR TITLE
fix(lib/grandpa): fix grandpa stall and various bugs

### PR DESCRIPTION
**Pull request by noot**
## Changes

- remove grandpa pause during catch-up functionality for now, as catch-up is buggy and will be fixed in a follow-up
- store greatest finalized round and set ID in database in specific key, previously was using (0,0) to designate this but wasn't always correct

## Tests

```
go test ./dot/state
go test ./lib/grandpa -short
```

## Issues

- related to #592

### CODE REVIEW

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: dot/state/block_finalisation.go [hunk 1]
```diff
@@ -94,23 +150,32 @@ func (bs *BlockState) SetFinalisedHash(hash common.Hash, round, setID uint64) er
...
 
-		logger.Trace("pruned block", "hash", rem, "number", header.Number)
-		bs.pruneKeyCh <- header
+		logger.Trace("pruned block", "hash", hash, "number", header.Number)
+		go func(header *types.Header) {
+			bs.pruneKeyCh <- header
+		}(header)
 	}
 
 	bs.lastFinalised = hash
```
[the remark this answers is not in the document]
Comment (noot) [reply dot/state/block_finalisation.go:172]: we don't want it to set `lastFinaliasedHash` if there's some error, and also this function is atomic due to the locking, so doesn't make a difference I don't think

File: dot/state/block_finalisation.go [hunk 2]
```diff
@@ -69,8 +75,58 @@ func (bs *BlockState) GetFinalisedHash(round, setID uint64) (common.Hash, error)
...
+	if err != nil {
+		return err
+	}
+
+	// higher setID takes precedence over round
+	if setID < currSetID || setID == currSetID && round <= currRound {
+		return nil
+	}
+
+	return bs.db.Put(highestRoundAndSetIDKey, roundSetIDKey(round, setID))
```
[the remark this answers is not in the document]
Comment (noot) [reply dot/state/block_finalisation.go:89]: updated!

File: dot/state/block_finalisation.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (EclesioMeloJunior) [reply dot/state/block_finalisation.go:172]: got it!

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: lib/grandpa/message_tracker.go [hunk 3]
```diff
@@ -69,6 +69,7 @@ func (t *tracker) add(v *networkVoteMessage) {
 	}
 
 	t.mapLock.Lock()
+	// TODO: change to map of maps, this allows duplicates
```
Comment (EclesioMeloJunior) [inline lib/grandpa/message_tracker.go:72]: created issue #1718 for this TODO

File: lib/grandpa/grandpa.go [hunk 4]
```diff
@@ -345,7 +365,7 @@ func (s *Service) initiate() error {
 		}
 
 		if err != nil {
-			return err
+			continue
```
[the remark this answers is not in the document]
Comment (noot) [reply lib/grandpa/grandpa.go:368]: added!

**Review by EclesioMeloJunior — APPROVED** [review]

**Pull request merged**